### PR TITLE
Add oneflow.mode

### PR DESCRIPTION
**Pull request by youxiudeshouyeren**
当前进度
+ CPU上简单测试没问题
image

待完成
~- [ ] CUDA kernel开发 暂时搁置，等thrust支持后加入cuda实现~
- [x] Docstr编写
- [x] test编写

### CODE REVIEW

File: oneflow/ir/include/OneFlow/OneFlowUserOps.td [hunk 1]
```diff
@@ -4126,6 +4126,20 @@ def OneFlow_SearchSortedScalarOp : OneFlow_BaseOp<"searchsorted_scalar", [NoSide
...
+  let output = (outs
+    OneFlow_Tensor:$values,
+    OneFlow_Tensor:$indices
+  );
+  let has_logical_tensor_desc_infer_fn = 1;
+  let has_physical_tensor_desc_infer_fn = 1;
+  let has_get_sbp_fn = 1;
+  let has_data_type_infer_fn = 1;
+}
+
```
Comment (marigoold) [inline oneflow/ir/include/OneFlow/OneFlowUserOps.td:4142]: 这里应该有个 dim 和 keepdim 的 attr

Comment (youxiudeshouyeren) [reply oneflow/ir/include/OneFlow/OneFlowUserOps.td:4142]: dim和keepdim在functor层面处理了 这里应该不用向kernel传递过来了吧

Comment (marigoold) [reply oneflow/ir/include/OneFlow/OneFlowUserOps.td:4142]: > dim和keepdim在functor层面处理了 这里应该不用向kernel传递过来了吧

如果能在 kernel 里面都做完的话，尽量在 kernel 里面做，functor 调用 kernel 有额外开销

Comment (marigoold) [reply oneflow/ir/include/OneFlow/OneFlowUserOps.td:4142]: > dim和keepdim在functor层面处理了 这里应该不用向kernel传递过来了吧

可以参考下有类似参数的 op 是怎么处理的，比如 std var 之类

File: python/oneflow/test/modules/test_mode.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        return torch.mode(x)
+
+    @autotest(n=5, auto_backward=False, check_graph=False)
+    def test_mode_reduce_one_dim_0size(test_case):
+        device = cpu_device()
+        x = random_tensor(ndim=3, dim1=0, requires_grad=False).to(device)
+        return torch.mode(x, 0)
+
+    @autotest(n=5, auto_backward=False)
+    def test_mode_return_type(test_case):
```
Comment (daquexian) [inline python/oneflow/test/modules/test_mode.py:53]: 这个测试是不是没有必要，其它几个测试会自动检查 named tuple 的 field name 是否一致

File: python/oneflow/test/modules/test_mode.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        x = random_tensor(3, 4)
+        result = x.mode(1)
+        return result.values, result.indices
+
```
Comment (daquexian) [inline python/oneflow/test/modules/test_mode.py:57]: 可以再加一个 profile 函数

File: python/oneflow/test/modules/test_mode.py (hunk 2 above)
Comment (youxiudeshouyeren) [reply python/oneflow/test/modules/test_mode.py:53]: 好的 我把它删除

File: python/oneflow/test/modules/test_mode.py (hunk 3 above)
Comment (youxiudeshouyeren) [reply python/oneflow/test/modules/test_mode.py:57]: profile是对cpu和cuda两个设备测量的，这个算子目前只有cpu实现，用profile会报错

Comment (daquexian) [reply python/oneflow/test/modules/test_mode.py:57]: 好的，那先不加了

File: python/oneflow/test/modules/test_mode.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+import unittest
+from oneflow.test_utils.automated_test_util import *
+import oneflow as flow
+import oneflow.unittest
+
+
```
Comment (doombeaker) [inline python/oneflow/test/modules/test_mode.py:21]: 我有个关于 `torch.mode` 的行为不是很明白的地方，没有去看源码确认。想偷懒在这里请教下。就是我自己写的例子，发现当 `dim` 值设置不一样时，返回的 indices 的行为是变化的。

比如，对于2维的张量，`dim=1` 时，`indices` 是返回较大者：

```
>>> a = torch.tensor([[1, 2, 1, 3, 5],[3,3,4,9,7], [11, 12, 15, 15, 19]])
>>> a
tensor([[ 1,  2,  1,  3,  5],
        [ 3,  3,  4,  9,  7],
        [11, 12, 15, 15, 19]])
>>> torch.mode(a,1)
torch.return_types.mode(
values=tensor([ 1,  3, 15]),
indices=tensor([2, 1, 3]))  # 不是 0, 0, 2
```

而对于以下的张量， `dim=0` 时，`indices` 是返回较小的 `indices`。

```python
>>> b =  torch.tensor([[  5,   3,   9,   9,  11],
...         [  8,   6,  12,  12,  14],
...         [ -4,  -6,   0,   0,   2],
...         [ 11,   9,  15,  15,  17],
...         [ -7,  -9,  -3,  -3,  -1],
...         [ 11,   9,  15,  15,  17],
...         [  3,   1,   7,   7,   9],
...         [  1,  -1,   5,   5,   7],
...         [ -5,  -7,  -1,  -1,   1],
...         [ -6,  -8,  -2,  -2,   0],
...         [  7,   5,  11,  11,  13],
...         [ 12,  10,  16,  16,  18],
...         [ -2,  -4,   2,   2,   4],
...         [  4,   2,   8,   8,  10],
...         [  2,   0,   6,   6,   8],
...         [ -1,  -3,   3,   3,   5],
...         [  3,   1,   7,   7,   9],
...         [  7,   5,  11,  11,  13],
...         [ 16,  14,  20,  20,  22],
...         [ -2,  -4,   2,   2,   4],
...         [ -2,  -4,   2,   2,   4],
...         [ -3,  -5,   1,   1,   3],
...         [ 10,   8,  14,  14,  16],
...         [  5,   3,   9,   9,  11],
...         [ -2,  -4,   2,   2,   4],
...         [ -2,  -4,   2,   2,   4],
...         [ 12,  10,  16,  16,  18],
...         [  9,   7,  13,  13,  15],
...         [  9,   7,  13,  13,  15],
...         [ -5,  -7,  -1,  -1,   1],
...         [ -9, -11,  -5,  -5,  -3],
...         [  7,   5,  11,  11,  13],
...         [  6,   4,  10,  10,  12],
...         [  5,   3,   9,   9,  11],
...         [ -1,  -3,   3,   3,   5],
...         [  3,   1,   7,   7,   9],
...         [  3,   1,   7,   7,   9],
...         [  6,   4,  10,  10,  12],
...         [  9,   7,  13,  13,  15],
...         [ -8, -10,  -4,  -4,  -2],
...         [  5,   3,   9,   9,  11],
...         [ -6,  -8,  -2,  -2,   0],
...         [  4,   2,   8,   8,  10],
...         [  3,   1,   7,   7,   9],
...         [ 10,   8,  14,  14,  16],
...         [  0,  -2,   4,   4,   6],
...         [  7,   5,  11,  11,  13],
...         [  2,   0,   6,   6,   8],
...         [ -4,  -6,   0,   0,   2],
...         [  9,   7,  13,  13,  15]])
>>> b
tensor([[  5,   3,   9,   9,  11],
        [  8,   6,  12,  12,  14],
        [ -4,  -6,   0,   0,   2],
        [ 11,   9,  15,  15,  17],
        [ -7,  -9,  -3,  -3,  -1],
        [ 11,   9,  15,  15,  17],
        [  3,   1,   7,   7,   9],
        [  1,  -1,   5,   5,   7],
        [ -5,  -7,  -1,  -1,   1],
        [ -6,  -8,  -2,  -2,   0],
        [  7,   5,  11,  11,  13],
        [ 12,  10,  16,  16,  18],
        [ -2,  -4,   2,   2,   4],
        [  4,   2,   8,   8,  10],
        [  2,   0,   6,   6,   8],
        [ -1,  -3,   3,   3,   5],
        [  3,   1,   7,   7,   9],
        [  7,   5,  11,  11,  13],
        [ 16,  14,  20,  20,  22],
        [ -2,  -4,   2,   2,   4],
        [ -2,  -4,   2,   2,   4],
        [ -3,  -5,   1,   1,   3],
        [ 10,   8,  14,  14,  16],
        [  5,   3,   9,   9,  11],
        [ -2,  -4,   2,   2,   4],
        [ -2,  -4,   2,   2,   4],
        [ 12,  10,  16,  16,  18],
        [  9,   7,  13,  13,  15],
        [  9,   7,  13,  13,  15],
        [ -5,  -7,  -1,  -1,   1],
        [ -9, -11,  -5,  -5,  -3],
        [  7,   5,  11,  11,  13],
        [  6,   4,  10,  10,  12],
        [  5,   3,   9,   9,  11],
        [ -1,  -3,   3,   3,   5],
        [  3,   1,   7,   7,   9],
        [  3,   1,   7,   7,   9],
        [  6,   4,  10,  10,  12],
        [  9,   7,  13,  13,  15],
        [ -8, -10,  -4,  -4,  -2],
        [  5,   3,   9,   9,  11],
        [ -6,  -8,  -2,  -2,   0],
        [  4,   2,   8,   8,  10],
        [  3,   1,   7,   7,   9],
        [ 10,   8,  14,  14,  16],
        [  0,  -2,   4,   4,   6],
        [  7,   5,  11,  11,  13],
        [  2,   0,   6,   6,   8],
        [ -4,  -6,   0,   0,   2],
        [  9,   7,  13,  13,  15]])
>>> torch.mode(b,0)
torch.return_types.mode(
values=tensor([-2, -4,  2,  2,  4]),
indices=tensor([12, 12, 12, 12, 12])) # 不是 [25, 25, 25, 25, 25]
```

我是偶然发现上面的问题的，奇怪的是，我如果把以上 `b` 的例子变小, `indices` 又会取较大的：

```pyhon
>>> b =  torch.tensor([[  5,   3,   9,   9,  11],
...             [  3,   1,   7,   7,   9],
...             [  1,  -1,   5,   5,   7],
...             [ -2,  -4,   2,   2,   4],
...             [  3,   1,   7,   7,   9],
...             [  7,   5,  11,  11,  13],
...             [ -2,  -4,   2,   2,   4],
...             [ -2,  -4,   2,   2,   4],
...             [ 12,  10,  16,  16,  18],
...             [  9,   7,  13,  13,  15]])
>>> b
tensor([[ 5,  3,  9,  9, 11],
        [ 3,  1,  7,  7,  9],
        [ 1, -1,  5,  5,  7],
        [-2, -4,  2,  2,  4],
        [ 3,  1,  7,  7,  9],
        [ 7,  5, 11, 11, 13],
        [-2, -4,  2,  2,  4],
        [-2, -4,  2,  2,  4],
        [12, 10, 16, 16, 18],
        [ 9,  7, 13, 13, 15]])
>>> torch.mode(b, 0)
torch.return_types.mode(
values=tensor([-2, -4,  2,  2,  4]),
indices=tensor([7, 7, 7, 7, 7])) # 不是 [3, 3, 3, 3, 3]
```

我不知道为什么会有以上这么奇怪的行为……
还是我没正确理解相关的内涵，犯了低级的错误……

我想确认的是：
1. 以上“奇怪的行为“， oneflow 和 torch 是对齐的吗
2. 有可能把以上的不同行为，加入 test_case 中吗（或者其实已经包含了？）

File: oneflow/user/kernels/mode_kernel.cpp [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    const user_op::Tensor* in = ctx->Tensor4ArgNameAndIndex("input", 0);
+    const int64_t num_axes = in->shape_view().NumAxes();
+    const int64_t size = in->shape_view().elem_cnt();
+    if (size == 0) return;
+    const int64_t stride = in->shape_view().At(num_axes - 1);
+    const int64_t instance_num = size / stride;
+    user_op::Tensor* values = ctx->Tensor4ArgNameAndIndex("values", 0);
+    user_op::Tensor* indices = ctx->Tensor4ArgNameAndIndex("indices", 0);
+    user_op::Tensor* tmp_buffer = ctx->Tensor4ArgNameAndIndex("tmp_buffer", 0);
+    Memcpy<DeviceType::kCPU>(ctx->stream(), tmp_buffer->mut_dptr<void>(), in->dptr<void>(),
```
Comment (doombeaker) [inline oneflow/user/kernels/mode_kernel.cpp:39]: 我印象中 `Memcpy<DeviceType>` 这个模板只考虑了 `kCUDA` 和 `kCPU`，未来会被 Primitive 抽象层中的方法替代。
所以这里可能是可以换成 `NewMemcpyPrimitive` 方法，参考:
https://github.com/Oneflow-Inc/oneflow/blob/master/oneflow/user/kernels/matmul_kernels.cpp#L173-L176

不过我不确认，如果看过以上的参考还不确定的话，可以找晓雨或者俊丞确认下，是否用 `NewMemcpyPrimitive` 更好。

File: python/oneflow/test/modules/test_mode.py (hunk 4 above)
Comment (youxiudeshouyeren) [reply python/oneflow/test/modules/test_mode.py:21]: 1. 将上面的例子分别用oneflow和torch跑了一遍，结果是对齐的
2. 以上行为应该是包含在test_case中了，因为对不同大小、不同dim、不同参数都有测试

我的写法是借鉴的torch的写法，查阅了下源码，问题在std::sort上。
由于要返回indices，所以对value和indices组成的pair进行了排序，这里将value作为排序时的key。而std::sort是不稳定的排序，主要采用了快速排序算法，所以排序时不能保证相同key的pair的先后顺序在排序后还是具有之前的先后顺序。如果要始终保持返回较大的indices，可以更换成std::stable_sort，但是要保持对齐torch的话，就还是用std::sort了。

File: oneflow/core/autograd/gradient_funcs/mode.cpp [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    }
+    return Maybe<void>::Ok();
+  }
+  Maybe<void> Apply(const ModeCaptureState* ctx, const TensorTuple& out_grads,
+                    TensorTuple* in_grads) const override {
+    if (ctx->requires_grad) {
+      in_grads->resize(1);
+      const auto& input = JUST(VectorAt(ctx->SavedTensors(), 0));
+      const auto& indices = JUST(functional::Unsqueeze(JUST(VectorAt(ctx->SavedTensors(), 1)), -1));
+      const auto& dout = JUST(functional::Unsqueeze(JUST(VectorAt(out_grads, 0)), -1));
```
[the remark this answers is not in the document]
Comment (youxiudeshouyeren) [reply oneflow/core/autograd/gradient_funcs/mode.cpp:49]: 这里求梯度的话，我觉得梯度的维度应该要和input的维度保持一致，所以就始终keepdim了。我查阅了下其他类似的算子，比如median求中位数，也是这样做的，所以应该不需要做判断吧

File: oneflow/user/kernels/mode_kernel.cpp (hunk 5 above)
Comment (youxiudeshouyeren) [reply oneflow/user/kernels/mode_kernel.cpp]: 已更新到NewMemcpyPrimitive的方法

File: python/oneflow/test/modules/test_mode.py (hunk 4 above)
Comment (doombeaker) [reply python/oneflow/test/modules/test_mode.py:21]: 好的，学习了

File: oneflow/core/autograd/gradient_funcs/mode.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (doombeaker) [reply oneflow/core/autograd/gradient_funcs/mode.cpp:49]: 哦哦，我看了下。其实是因为你的 keepdim 是在 functor 层中处理的，所以这里不需要判断了。
我没有意见。

**Review by doombeaker — APPROVED** [review]

**Review by marigoold — APPROVED** [review]
